File: Binary Search the Answer (int).cpp
#include <bits/stdc++.h>
using namespace std;
template <typename T, typename Can> T bsearch(T L, T R, const Can& can, bool left_feasible = true) {
    static_assert(is_convertible<decltype(can), function<bool(T)>>::value);
    T& feasible = left_feasible ? L : R;
    T& infeasible = left_feasible ? R : L;
    while(R - L > 1) {
        T M = L + (R - L) / 2;
        (can(M) ? feasible : infeasible) = M;
    }
    return feasible;
}

File: Binary Search the Answer (int).py
def bsearch(L, R, can, left_feasible = True):
    while R - L > 1:
        M = L + (R - L) // 2
        if can(M) ^ left_feasible:
            R = M
        else:
            L = M
    return L if left_feasible else R